be/src/vec/exec/join/process_hash_table_probe_impl.h

```diff
@@ -674,20 +675,38 @@ Status ProcessHashTableProbe<JoinOpType>::process_data_in_hashtable(HashTableTyp
             }
         };
 
-        for (; iter != hash_table_ctx.hash_table.end() && block_size < _batch_size; ++iter) {
+        for (; iter != hash_table_ctx.hash_table.end() && block_size < _batch_size;) {
             auto& mapped = iter->get_second();
             if constexpr (std::is_same_v<Mapped, RowRefListWithFlag>) {
                 if (mapped.visited) {
-                    for (auto it = mapped.begin(); it.ok(); ++it) {
-                        if constexpr (JoinOpType == TJoinOp::RIGHT_SEMI_JOIN) {
+                    if constexpr (JoinOpType == TJoinOp::RIGHT_SEMI_JOIN) {
+                        for (auto it = mapped.begin(); it.ok(); ++it) {
                             insert_from_hash_table(it->block_offset, it->row_num);
                         }
+                        ++iter;
+                    } else {
+                        for (; visited_iter.ok() && block_size < _batch_size; ++visited_iter) {
+                            insert_from_hash_table(visited_iter->block_offset,
+                                                   visited_iter->row_num);
+                        }
+                        if (!visited_iter.ok()) {
+                            ++iter;
+                        }
                     }
                 } else {
-                    for (auto it = mapped.begin(); it.ok(); ++it) {
-                        if constexpr (JoinOpType != TJoinOp::RIGHT_SEMI_JOIN) {
-                            insert_from_hash_table(it->block_offset, it->row_num);
+                    if constexpr (JoinOpType != TJoinOp::RIGHT_SEMI_JOIN) {
+                        visited_iter = mapped.begin();
+                        for (; visited_iter.ok() && block_size < _batch_size; ++visited_iter) {
+                            insert_from_hash_table(visited_iter->block_offset,
+                                                   visited_iter->row_num);
+                        }
+                        if (visited_iter.ok()) {
+                            mapped.visited = true;
```

Comment:
why here change the `mapped.visited`

Reply:
After `mapped.visited`, next time can run `process_hash_table_probe_impl.h:681` `if (mapped.visited)`, continue to read the current mapped

Reply:
Do you mean to judge whether to continue reading based on `visited_iter.ok()`

Reply:
fixed